map_model/src/make/traffic_signals.rs

```diff
@@ -329,6 +332,426 @@ fn stage_per_road(map: &Map, i: IntersectionID) -> ControlTrafficSignal {
+    fn make_crosswalk_variable(ts: &mut ControlTrafficSignal) {
+        const MIN_CROSSWALK_TIME: Duration = Duration::const_seconds(15.0);
+        for mut s in ts.stages.iter_mut() {
+            if let Some(mut duration) = s.contains_only_crosswalks(&ts.movements) {
+                match s.stage_type {
+                    StageType::Fixed(_) => {
+                        // get the minimum duration, but never less than 15 seconds
+                        if duration < MIN_CROSSWALK_TIME {
+                            duration = MIN_CROSSWALK_TIME;
+                        }
+                        s.stage_type = StageType::Variable(
+                            duration,
+                            Duration::const_seconds(1.0),
+                            Duration::const_seconds(1.0),
+                        )
+                    }
+                    _ => (),
+                }
+            }
+        }
+    }
+
+    fn merge_stages(ts: &ControlTrafficSignal) -> Option<ControlTrafficSignal> {
+        for s_src in &ts.stages {
+            // s_src is the stage we want to apply to the other stages
+            for s_dst in &ts.stages {
+                if s_src == s_dst {
+                    continue;
+                }
+                let mut merged_stage = s_dst.clone();
+                for m in &s_src.protected_movements {
+                    merged_stage.protected_movements.insert(m.clone());
+                }
+                let mut maybe_ts = ts.clone();
+                // insert at the head, keeping crosswalk last
+                maybe_ts.stages.insert(0, merged_stage);
+                if maybe_ts.validate().is_ok() {
+                    println!("can merge s_src={:#?} and s_dst={:#?}", s_src, s_dst);
+                    let mut stages: Vec<Stage> = Vec::new();
+                    for s in maybe_ts.stages {
+                        if s != *s_src && s != *s_dst {
+                            stages.push(s);
+                        }
+                    }
+                    maybe_ts.stages = stages;
+                    return Some(maybe_ts);
+                }
+            }
+        }
+        None
+    }
+
+    // sometimes protected oncoming left turns aren't possible.
+    fn is_conflict(ts: &ControlTrafficSignal, stage: &Stage) -> Option<(MovementID, MovementID)> {
+        for m1 in stage.protected_movements.iter().map(|m| &ts.movements[m]) {
+            for m2 in stage.protected_movements.iter().map(|m| &ts.movements[m]) {
+                if m1.conflicts_with(m2) {
+                    return Some((m1.id, m2.id));
+                }
+            }
+        }
+        return None;
+    }
+
+    fn protected_yield_stage(p: MovementID, y: MovementID) -> Stage {
+        let mut stage = Stage::new();
+        stage.protected_movements.insert(p);
+        stage.yield_movements.insert(y);
+        stage
+    }
+
+    fn multi_way_stages(map: &Map, id: IntersectionID) -> Option<ControlTrafficSignal> {
```

Comment:
A docstring would be helpful here; this one is a little complicated to follow. I wish it was easier to just embed pictures in docs; following an example would make things like this incredibly easy to understand.